File: rust/src/bin/p0053.rs
// Project Euler: Problem 53

euler::run_solver!(53);

fn solve() -> String {
    compute(100, 1_000_000).to_string()
}

fn compute(max_n: u64, thr: u64) -> u64 {
    debug_assert!(max_n > 0);

    // find the first central binomial coefficient C(2k, k) which is larger than 'thr'.
    let mut c: u64 = 2;
    let mut x: u64 = 1;
    while c <= thr {
        x += 1;
        c = c * (4 * x - 2) / x;
    }

    // the start position 'c' = C(n, r) is C(2k-1, k-1). And then set x = (n - r + 1).
    let mut r: u64 = x - 1;
    x += 1;
    c /= 2;

    let mut ans: u64 = 0;
    for n in (r + x - 1)..=max_n {
        while c > thr {
            c = c * r / x;
            r -= 1;
            x += 1;
        }
        ans += n - 2 * r - 1;

        // go to next row (down-left)
        c = c * (n + 1) / x;
        x += 1;
    }

    ans
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0053_23_1000000() {
        assert_eq!(compute(23, 1_000_000), 4);
    }

    #[test]
    fn p0053_100_1000000() {
        assert_eq!(compute(100, 1_000_000), 4075);
    }
}

File: ocaml/src/p0053.ml
(* Project Euler: Problem 53 *)

(*
 *  >>> import math
 *  >>> math.log10(math.factorial(100))
 *  157.97000365471578
 *  >>> math.log10(math.factorial(50))
 *  64.48307487247203
 *
 *  Since the above result, I need to use arbitrary-precision arithmetic library or another method.
 *  Therefore, I use Pascal's triangle formula.
 *  C(n,r) = C(n-1,r-1) + C(n-1,r)
 *  C(n,0) = C(n,n) = 1
 *
 *  C(2n, n) = C(2(n-1), n-1) * (3 + (n-2)/n)
 *           = C(2(n-1), n-1) * (4n-2)/n
 *)

open Core

let compute max_n thr =
  assert (max_n > 0);

  (* find the first central binomial coefficient C(2k, k) which is larger than 'thr'. *)
  let c = ref 2
  and x = ref 1 in
  while !c <= thr do
    x := !x + 1;
    c := !c * ((4 * !x) - 2) / !x
  done;

  (* the start position 'c' = C(n, r) is C(2k-1, k-1). And then set x = (n - r + 1). *)
  let r = ref (!x - 1) in
  x := !x + 1;
  c := !c / 2;

  let rec loop ans = function
    | [] -> ans
    | n :: ns ->
      while !c > thr do
        c := !c * !r / !x;
        r := !r - 1;
        x := !x + 1
      done;

      (* go to next row (down-left) *)
      c := !c * (n + 1) / !x;
      x := !x + 1;
      loop (ans + n - (2 * !r) - 1) ns
  in
  loop 0 (List.range (!r + !x - 1) max_n ~stop:`inclusive)
;;

let solve () = compute 100 1_000_000 |> Int.to_string

(* Test *)

let%test_unit "23, 1_000_000" = [%test_eq: int] (compute 23 1_000_000) 4
let%test_unit "100, 1_000_000" = [%test_eq: int] (compute 100 1_000_000) 4075
